### PR TITLE
[Fix] Swap package dependency ruff for black and autoflake

### PR DESCRIPTION
**Pull request by maxschulz-COL**
## Description
Given our continued difficulties to make `ruff` run on WebAssembly, we decide to swap the usage of ruff via subprocess for using black and autoflake. Using ruff is still the preferred goal, but as long as there is no python API, as well as a WASM compiled version available, we will stick with the current solution.

Note:
ruff was previous used in 2 places for Vizro-AI code string formatting:

inside vizro-ai plot
inside vizro-core pydantic to python (vizro models to python code) conversion.
This PR is mainly to drop the ruff use for these two.

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-ai/changelog.d/20240930_141508_maximilian_schulz_swap_ruff_for_black.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+-->
+<!--
+### Added
+
+- A bullet item for the Added category with a link to the relevant PR at the end of your entry, e.g. Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))
+
+-->
+<!--
+### Changed
```
Comment (maxschulz-COL) [inline vizro-ai/changelog.d/20240930_141508_maximilian_schulz_swap_ruff_for_black.md:26]: Add changlog

**Review by lingyielia — APPROVED** [review]

**Review by huong-li-nguyen — COMMENTED** [review]
Not sure if I fully understand the changes - is our tool to replace ruff completely or just in the places where we had to add it as a requirement? 

For example there are still some mentions of ruff and we use it e.g. for pydocstyle etc.
Does this not have to be replaced? Or is this fine because it's run via our pre-commit hook?
```

[tool.ruff]
line-length = 120

[tool.ruff.format]
docstring-code-format = true

[tool.ruff.lint.pydocstyle]
convention = "google"

[tool.ruff.lint.pylint]
max-args = 6

...
```

**Comment by lingyielia** [comment]
> Is our goal to replace ruff completely or just as a mkdocs requirement?
> 
> For example there are still some mentions of ruff and we use it e.g. for pydocstyle etc. Does this not have to be replaced? Or is this fine because it's run via our pre-commit hook?
> 
> ```
> 
> [tool.ruff]
> line-length = 120
> 
> [tool.ruff.format]
> docstring-code-format = true
> 
> [tool.ruff.lint.pydocstyle]
> convention = "google"
> 
> [tool.ruff.lint.pylint]
> max-args = 6
> 
> ...
> ```

ruff was previous used in 2 places for Vizro-AI code string formatting:
1. inside vizro-ai plot
2. inside vizro-core pydantic to python (vizro models to python code) conversion.

This is PR is mainly to drop the ruff use for these two.

**Review by huong-li-nguyen — APPROVED** [review]

File: vizro-core/changelog.d/20240930_141041_maximilian_schulz_swap_ruff_for_black.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+-->
+<!--
+### Changed
+
+- A bullet item for the Changed category with a link to the relevant PR at the end of your entry, e.g. Enable feature XXX ([#1](https://github.com/mckinsey/vizro/pull/1))
+
+-->
+
+### Deprecated
```
Comment (antonymilne) [inline vizro-core/changelog.d/20240930_141041_maximilian_schulz_swap_ruff_for_black.md:32]: This should be in changed, not deprecated.

File: vizro-ai/src/vizro_ai/plot/_response_models.py [hunk 3]
```diff
@@ -26,12 +27,10 @@
...
-    linted = subprocess.check_output(
-        ["ruff", "check", "--fix", "--exit-zero", "--silent", "--isolated", "-"], input=code_string, encoding="utf-8"
-    )
-    formatted = subprocess.check_output(
-        ["ruff", "format", "--silent", "--isolated", "-"], input=linted, encoding="utf-8"
-    )
+    # While we wait for the API, we can autoflake and black to process code strings.
+
+    removed_imports = autoflake.fix_code(code_string, remove_all_unused_imports=True)
+    formatted = format_str(removed_imports, mode=FileMode())
```
Comment (antonymilne) [inline vizro-ai/src/vizro_ai/plot/_response_models.py:33]:
```suggestion
    # Black doesn't yet have a Python API, so format_str might not work at some point in the future.
    # https://black.readthedocs.io/en/stable/faq.html#does-black-have-an-api
    formatted = black.format_str(removed_imports, mode=black.Mode())
```

* comment would be useful
* `FileMode` is a legacy name for `Mode`
* namespace `black` is a bit nicer than having two vaguely named global imports

File: vizro-core/src/vizro/models/_base.py [hunk 4]
```diff
@@ -73,12 +74,10 @@ def _patch_vizro_base_model_dict():
...
-    linted = subprocess.check_output(
-        ["ruff", "check", "--fix", "--exit-zero", "--silent", "--isolated", "-"], input=code_string, encoding="utf-8"
-    )
-    formatted = subprocess.check_output(
-        ["ruff", "format", "--silent", "--isolated", "-"], input=linted, encoding="utf-8"
-    )
+    # While we wait for the API, we can use autoflake and black to process code strings
+
+    removed_imports = autoflake.fix_code(code_string, remove_all_unused_imports=True)
+    formatted = format_str(removed_imports, mode=FileMode())
```
Comment (antonymilne) [inline vizro-core/src/vizro/models/_base.py:80]:
```suggestion
    # Black doesn't yet have a Python API, so format_str might not work at some point in the future.
    # https://black.readthedocs.io/en/stable/faq.html#does-black-have-an-api
    formatted = black.format_str(removed_imports, mode=black.Mode())
```

* comment would be useful
* `FileMode` is a legacy name for `Mode`
* namespace `black` is a bit nicer than having two vaguely named global imports

**Review by antonymilne — APPROVED** [review]
A sad but necessary change :(

Please could you give a quick example of how this output is different to what ruff gave before? I guess the fact that none of the unit tests were changed means it's pretty much the same?

**Comment by maxschulz-COL** [comment]
> A sad but necessary change :(
> 
> Please could you give a quick example of how this output is different to what ruff gave before? I guess the fact that none of the unit tests were changed means it's exactly the same for all our example cases?

Yes, the magic of unit tests :)

So in general not all code string would end up the same, as we are only formatting with black rules (and they may not 100% match - but seem to for the examples) and we only remove unused imports, and no other formats (which previously was the case for ruff but it looks like it was also unused)

**Comment by maxschulz-COL** [comment]
> Is our goal to replace ruff completely or just as a mkdocs requirement?

We are not removing our development dependency and tooling, but the package dependency.

**Pull request merged**